File: spec/functional_test/fixtures/scala/play/app/controllers/HomeController.scala
package controllers

import play.api.mvc._
import play.api.libs.json._

class HomeController extends BaseController {
  def index() = Action {
    Ok("Welcome to Play")
  }
}

class Users extends BaseController {
  def list() = Action {
    Ok("User list")
  }

  def show(id: Long) = Action {
    Ok(s"User $id")
  }

  def create() = Action {
    Ok("User created")
  }

  def update(id: Long) = Action {
    Ok(s"User $id updated")
  }

  def delete(id: Long) = Action {
    Ok(s"User $id deleted")
  }
}

class Search extends BaseController {
  def search(q: String, filter: String) = Action {
    Ok(s"Search results for $q with filter $filter")
  }
}

class Posts extends BaseController {
  def show(userId: Long, postId: Long) = Action {
    Ok(s"Post $postId from user $userId")
  }
}

class Items extends BaseController {
  def list(category: Option[String], page: Int) = Action {
    Ok(s"Items in category ${category.getOrElse("all")}, page $page")
  }
}

class Files extends BaseController {
  def download(path: String) = Action {
    Ok(s"Downloading file from $path")
  }
}

class Upload extends BaseController {
  def file() = Action {
    Ok("File uploaded")
  }
}

class Api extends BaseController {
  def protectedEndpoint() = Action { request =>
    val authToken = request.headers.get("Authorization").getOrElse("none")
    val sessionId = request.cookies.get("session_id").map(_.value).getOrElse("none")
    Ok(s"Protected endpoint - Auth: $authToken, Session: $sessionId")
  }

  def postData() = Action { request =>
    val contentType = request.headers.get("Content-Type").getOrElse("unknown")
    val json = request.body.asJson
    Ok(Json.toJson(json))
  }
}

class Assets extends BaseController {
  def at(path: String, file: String) = Action {
    Ok(s"Asset $path/$file")
  }
}

File: spec/functional_test/fixtures/java/play/app/controllers/HomeController.java
package controllers;

import play.mvc.Controller;
import play.mvc.Http;
import play.mvc.Result;
import play.libs.Json;
import com.fasterxml.jackson.databind.JsonNode;

public class HomeController extends Controller {
    public Result index() {
        return ok("Welcome to Play");
    }
}

class Users extends Controller {
    public Result list() {
        return ok("User list");
    }

    public Result show(Long id) {
        return ok("User " + id);
    }

    public Result create() {
        return ok("User created");
    }

    public Result update(Long id) {
        return ok("User " + id + " updated");
    }

    public Result delete(Long id) {
        return ok("User " + id + " deleted");
    }
}

class Search extends Controller {
    public Result search(String q, String filter) {
        return ok("Search results for " + q + " with filter " + filter);
    }
}

class Posts extends Controller {
    public Result show(Long userId, Long postId) {
        return ok("Post " + postId + " from user " + userId);
    }
}

class Items extends Controller {
    public Result list(String category, Integer page) {
        return ok("Items in category " + category + ", page " + page);
    }
}

class Files extends Controller {
    public Result download(String path) {
        return ok("Downloading file from " + path);
    }
}

class Upload extends Controller {
    public Result file() {
        return ok("File uploaded");
    }
}

class Api extends Controller {
    private static final String TRACE_HEADER = "X-Trace";
    private static final String SESSION_COOKIE = "session_id";

    public Result protectedEndpoint() {
        String authToken = request().header("Authorization");
        Http.Cookie sessionCookie = request().cookie("session_id");
        String sessionId = sessionCookie != null ? sessionCookie.value() : "none";
        return ok("Protected endpoint - Auth: " + (authToken != null ? authToken : "none") + ", Session: " + sessionId);
    }

    public Result postData() {
        String contentType = request().header("Content-Type");
        JsonNode json = request().body().asJson();
        return ok(Json.toJson(json));
    }

    public Result modern(Http.Request request) {
        String trace = request.header(TRACE_HEADER).orElse("none");
        Http.Cookie sessionCookie = request.cookies().get(SESSION_COOKIE);
        JsonNode json = request.body().asJson();
        return ok(Json.toJson(json));
    }

    public static Result staticPost() {
        String header = request().header("X-Static");
        String body = request().body().asText();
        return ok(header + body);
    }
}

class Assets extends Controller {
    public Result at(String path, String file) {
        return ok("Asset " + path + "/" + file);
    }
}

class AsyncApi extends Controller {
    public java.util.concurrent.CompletionStage< Result > async(Http.Request request) {
        String trace = request.header("X-Async").orElse("none");
        return java.util.concurrent.CompletableFuture.completedFuture(ok(trace));
    }

    public play.libs.F.Promise<Result> legacy() {
        String legacy = request().header("X-Legacy-Async");
        return null;
    }
}

class ApiReports extends Controller {
    public Result show(Long reportId) {
        return ok("Report " + reportId);
    }

    public Result create(Http.Request request) {
        JsonNode json = request.body().asJson();
        return ok(Json.toJson(json));
    }
}
